Add some support for WPT tests in an Android emulator through WebDriver

SimonSapin opened the pull request:
This succeeds on my machine:

`./mach test-wpt --product servodriver --binary etc/run_in_headless_android_emulator.py --binary-arg servo-x86 --binary-arg target/i686-linux-android/release/servo.apk /_mozilla/mozilla/DOMParser.html`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/21213)
<!-- Reviewable:end -->

SimonSapin committed:
WebDriver timeout settings are not optional.

Not setting one of them in a SetTimeouts requests should not remove it.
This fixes a panicking `unwrap()`.

SimonSapin committed:
Fix a webdriver timeout during server start up

When sending a webdriver load URL command soon enough after starting Servo,
that command could time out with a logged warning:

```
constellation: Webdriver load for closed browsing context (0,2).
```

When `closed` in this case really meant not opened yet.

SimonSapin committed:
Add --binary-arg support in ./mach test-wpt --product servodriver

SimonSapin committed:
Fix Python errors in servodriver

SimonSapin committed:
Make servodriver wait until the server starts accepting TCP connections

SimonSapin committed:
Refactor run_in_headless_android_emulator.py into more functions

SimonSapin committed:
run_in_headless_android_emulator: add port forwarding for webdriver

SimonSapin committed:
run_in_headless_android_emulator: add support for user stylesheets

SimonSapin committed:
run_in_headless_android_emulator: add support for reverse port forwarding

SimonSapin committed:
run_in_headless_android_emulator: add some code comments

SimonSapin committed:
More Python fixes for servodriver

```
Traceback (most recent call last):
  File "/home/simon/servo1/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py", line 90, in run
    rv = commands[command](*args)
  File "/home/simon/servo1/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py", line 110, in wait
    self.executor.wait()
  File "/home/simon/servo1/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/base.py", line 209, in wait
    self.protocol.base.wait()
AttributeError: 'ServoWebDriverProtocol' object has no attribute 'base'
```

SimonSapin committed:
Don’t reverse-forward port "None"

```
 0:27.48 pid:4157 Traceback (most recent call last):
 0:27.48 pid:4157   File "/home/simon/servo1/etc/run_in_headless_android_emulator.py", line 212, in <module>
 0:27.48 pid:4157     sys.exit(main(*sys.argv[1:]))
 0:27.48 pid:4157   File "/home/simon/servo1/etc/run_in_headless_android_emulator.py", line 72, in main
 0:27.48 pid:4157     forward_webdriver(adb, args)
 0:27.48 pid:4157   File "/home/simon/servo1/etc/run_in_headless_android_emulator.py", line 163, in forward_webdriver
 0:27.48 pid:4157     ports = [int(part) for part in split if part]
 0:27.48 pid:4157 ValueError: invalid literal for int() with base 10: 'None'
```

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/webdriver_server/lib.rs, components/constellation/constellation.rs
 * @cbrewster: components/constellation/constellation.rs
 * @jgraham: components/webdriver_server/lib.rs
 * @paulrouget: components/constellation/constellation.rs

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/74.

jdm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit cdd64604cf0c3f837f2b20faa786d6d2595c22d3 with merge f9eeb92fe489cecf1268232402653e75ecfa84ad...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4740)

SimonSapin committed:
servodriver: fix setting preferences

This makes `/_mozilla/mozilla/webgl/context_creation_error.html` pass,
for example.

```
./ mach test-wpt --product servodriver \
  --binary etc/run_in_headless_android_emulator.py \
  --binary-arg servo-x86 \
  --binary-arg target/i686-linux-android/release/servo.apk \
  /_mozilla/mozilla/webgl/context_creation_error.html
```

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#74.

SimonSapin committed:
Add ./mach test-wpt-android

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#74.

SimonSapin commented:
@bors-servo try

This time with a couple WPT tests running on Android on CI.

bors-servo commented:
:hourglass: Trying commit a57647aa4d6f9470e9ba77ec529cba334fbdc0bf with merge 41a6a6a6b73485d07caffe89b9d99e2d48894da3...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9176)

SimonSapin commented:
> with a couple WPT tests running on Android on CI.

This new job failed on the CI infa :(

```
Running 2 tests in web-platform-tests

Unsupported test type wdspec for product servodriver
  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/DOMParser.html

  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/webgl/context_creation_error.html


Ran 2 tests finished in 128.0 seconds.
  • 0 ran as expected. 0 tests skipped.
  • 2 tests timed out unexpectedly
```

On my machine that same command works. Running on arbitrary directory with 141 tests finishes in 18 minutes:

```
$ ./mach test-wpt-android /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/ --release
Assuming --target i686-linux-android
Running 141 tests in web-platform-tests

Unsupported test type wdspec for product servodriver
  ▶ OK [expected TIMEOUT] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/acos/acos_001_to_006.html

  ▶ OK [expected TIMEOUT] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/asin/asin_001_to_006.html

  ▶ Unexpected subtest result in /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/asin/asin_001_to_006.html:
  └ PASS [expected NOTRUN] Overall test

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/build/build_177_to_178.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/distance/distance_001_to_006.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/exp2/exp2_009_to_012.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/functions/functions_033_to_040.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/functions/functions_081_to_088.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/functions/functions_097_to_104.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/functions/functions_105_to_112.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/lessThan/lessThan_001_to_008.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/log/log_009_to_012.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/mat3/mat3_001_to_006.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/notEqual/notEqual_001_to_008.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/pow/pow_017_to_024.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/step/step_001_to_006.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/struct/struct_033_to_040.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/swizzlers/swizzlers_049_to_056.html
  └   → ''

  ▶ ERROR [expected OK] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/swizzlers/swizzlers_113_to_120.html
  └   → ''

Ran 141 tests finished in 1090.0 seconds.
  • 123 ran as expected. 0 tests skipped.
  • 16 tests had errors unexpectedly
  • 2 tests unexpectedly okay
  • 1 tests had unexpected subtest results
```

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#74.

SimonSapin commented:
Removing that last commit (diff below) for now so that the rest of the PR can land.

```diff
diff --git a/etc/ci/buildbot_steps.yml b/etc/ci/buildbot_steps.yml
index 12d5a318d0..4cc1fd6d92 100644
--- a/etc/ci/buildbot_steps.yml
+++ b/etc/ci/buildbot_steps.yml
@@ -167,6 +167,7 @@ android-x86:
   - ./etc/ci/bootstrap-android-and-accept-licences.sh
   - env --unset ANDROID_NDK --unset ANDROID_SDK ./mach build --target i686-linux-android --release
   - env --unset ANDROID_NDK --unset ANDROID_SDK ./mach test-android-startup --release
+  - env --unset ANDROID_NDK --unset ANDROID_SDK ./mach test-wpt-android --release /_mozilla/mozilla/DOMParser.html /_mozilla/mozilla/webgl/context_creation_error.html 
   - bash ./etc/ci/lockfile_changed.sh
 
 android-nightly:
```

SimonSapin commented:
CC @jgraham, if you want to have a look as the wptrunner changes

jgraham commented on tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/executorservodriver.py:
```diff
@@ -23,8 +26,62 @@ def do_delayed_imports():
     import webdriver
 
+    global ServoCommandExtensions
+    class ServoCommandExtensions(object):
```
Why is this inside `do_delayed_imports`?

SimonSapin replied:
Because it relies on `webdriver.client.command`, and `webdriver` is delayed.

jgraham commented on tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/executorservodriver.py:
```diff
@@ -23,8 +26,62 @@ def do_delayed_imports():
+        return value
+
+
+class ServoBaseProtocolPart(BaseProtocolPart):
```
So, a more correct implementation would put all the logic for connecting via WebDriver in here (the long term goal is that we are able to share the executors and just encapsulate browser differences in the Protocol objects, but we're not there yet). Having said that I don't mind if you wait to refactor in that way.

SimonSapin committed:
servodriver: increase browser’s init_timeout

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#74.

nox commented:
r=me for everything but the wptrunner parts.

bors-servo commented:
:hourglass: Testing commit 2b5bba6f12481acf315b6c55f3ac129dce373bd6 with merge 3d886aec1ce37d8a9c2aa3d822ee9c2cdad97132...

bors-servo commented:
:hourglass: Testing commit 2b5bba6f12481acf315b6c55f3ac129dce373bd6 with merge 37a73d3bc09a8a3f12e4e2d42850419208000ba8...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4819)

jdm commented:
@bors-servo retry
* #20734

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/11963), [android-x86](http://build.servo.org/builders/android-x86/builds/89), [arm32](http://build.servo.org/builders/arm32/builds/12179), [arm64](http://build.servo.org/builders/arm64/builds/12226), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12268), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9182), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9182), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12318), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4738), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4749), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9009), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9055), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4754), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7668) are reusable. Rebuilding only [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4819)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11963), [android-x86](http://build.servo.org/builders/android-x86/builds/89), [arm32](http://build.servo.org/builders/arm32/builds/12179), [arm64](http://build.servo.org/builders/arm64/builds/12226), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12268), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9182), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9182), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12318), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4738), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4749), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9009), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9055), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4820), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4754), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7668)
Approved by: nox+jgraham
Pushing 37a73d3bc09a8a3f12e4e2d42850419208000ba8 to master...
